Added Support for the Teensy 3.2, 3.5, 3.6, 4.0 and 4.1 MCUs

mjs513 opened the pull request:
Since I primarily use Teensy 3.x and 4.x mcu's (ARM based) - https://www.pjrc.com/teensy/techspecs.html - and I wanted to try out DCC we wound up adding support for that series of processors.  We followed you structure as close as possible by creating a new Time.h file for the Teensy.

Might be interesting to added support for other processors.

EDIT:  https://youtu.be/-pHElT-Mgrc and https://youtu.be/5nRWtq1nhvM.  Also tested with your web throttle and app.

habazut commented:
Ther is definitely stuff in there that can not be merged right away like  while(!Serial); and replacement of analogreadfast with ordinary reads. Much of it hidden by non-changes of whitespace.

Regards,
Harald.

mjs513 commented:
Hi Harald
Just went through all the functions with the compare and looks like analogReadFast is the only function that I forgot to change to make it either Teensy or Arduino.  I just updated the fix for that issue.

As for while(!Serial) it still should work for Arduino but if it is deleted it would still work on the Teensy.  

Why all the changes in the white space not sure.

R.
Mike

Asbelos reviewed (changes requested):
Thanks for this, it shows we are really close.
However there are some issues:
1) we are already well into an advanced upgrade for unoWifi and nanoEvery and this cuts across this update... some of the updated files will no longer exist and significant changes will clash. Please take a look at the nanoEvery2 branch which contains all the latest portability simplifications and speed accelerations.  
2) Please don't change the default config.. if you can't do Wifi, see how we ignore it for a uno.
3) please dont mess with blank spaces.. yes they need tidying but this hides real changes.
4) You did not change the #warning about supported board types.

mjs513 commented:
@Asbelos
I have great timing on the PR don't I. :) Anyway just took a quick look through the nanoEvery branch.  Just a cursory look through the code changes it is certainly do able to update the nanoEvery changes but would require slightly different approach in the motorDriver functions.  The timer changes does make it simpler.  As for speed changes don't seem to have the same issues as the Arduino boards.

Thought I fixed the warning for different board types.

Anyways, it might be a better idea to close this PR and wait for the upgrade to CommandStation-EX and then incorporate the necessary changes for the Teensy.  If you think this is a better approach we can close this PR and will modify the upgrade.

Thanks for considering this PR though.

Asbelos commented:
> @Asbelos
> I have great timing on the PR don't I. :) Anyway just took a quick look through the nanoEvery branch. Just a cursory look through the code changes it is certainly do able to update the nanoEvery changes but would require slightly different approach in the motorDriver functions. The timer changes does make it simpler. As for speed changes don't seem to have the same issues as the Arduino boards.
> 
> Thought I fixed the warning for different board types.
> 
> Anyways, it might be a better idea to close this PR and wait for the upgrade to CommandStation-EX and then incorporate the necessary changes for the Teensy. If you think this is a better approach we can close this PR and will modify the upgrade.
> 
> Thanks for considering this PR though.

My reading of the Teensy forums suggests the MotorDriver FASTPIN technique should work on a Teensy unchanged. 

We need the faster analogRead on a uno/mega because the default read takes longer than the gaps in the DCC waveform. However, I'm experimenting with putting the register changes for analogRead sample time into the DCCTimer code so its set once for the lifetime of the CS. After that all the rest of the code can do the standard analogRead and it's one less place to cause portability problems.

FrightRisk commented:
I like the idea of integrating this. I use a Teensy for one of my Halloween animatronics. Nothing else could really run that firware because of the need for fast graphics. While the Teensy lends itself to our idea to have our whole CS on a board idea (integrating an H-Bridge), in order for this to truly appeal to the masses instead of just the Engineers, we would need a carrier board. Perhaps even one that uses the Arduino expansion board layout. I don't know if we can get Teensy's with headers soldered, that is another issue. The last is cost and supply. Everything about the Uno is built around the expandability, decades long track record, and wealth of plug and play accessory boards. Nothing else has this. And with it not being open source, it is completely against our mission. We can't depend on a device that could stop production or be limited to one or two suppliers. So this could be an option, but not something we can hang our hat on.

mjs513 commented:
@FrightRisk 
Just to answer your questions on the Teensy:
1.  Yes you can get a Teensy with headers already soldered on either through amazon or direct with PJRC (cheaper here).
2.   There are several carrier boards out there for the Teensy.  But since I liked the idea of having a T4.1 (which has the SD Card builtin, ethernet builtin - just need the connector to attach - and the ability to solder a flash chip to the underneath plus it runs at 600Mhz and is a 32-bit processor) I designed and am using my own Arduino compatible carrier board for the T4.1:
![IMG-0318](https://user-images.githubusercontent.com/5366213/106469518-66922a00-646d-11eb-8c47-37eb1cf45b42.png)
I am lousy at soldering but it wasn't bad to do the soldering of the components.  The tests I made were with the carrier board and a Arduino Motor shield.
3.  There is also a USBHost connector attached where with the provided library you could attach a joystick or a game controller - think of the possibilities.
4.  There are a minimum of 3 H/W Serial ports so no need to play with software Serial.
5.  You could put a T4.0 on the carrier board but not all the break out pins would be available.  All the shield pins would be though.

While I know that the Teensy may not be for everyone it does have its advantages over the Uno/Mega.  I do love my Mega though.

FrightRisk commented:
@mjs513 Thanks Mike. I may ask you to write up something that I can convert to put on our website. I am thinking of a "case history" section, or maybe just a page on other command station options. I could use help with documentation and pictures. It is the step by step that takes a long time to write.

mjs513 commented:
@FrightRisk  - Not a problem.  Just tell me what you need, I will be glad to help. I know the step by step takes the time especially if you want me to show how to do the carrier board for the Teensy 4.1/4.0.  Once you have that its pretty much the same thing you would do for an UNO or a Mega.

jakespeed1311 commented:
Hello,
i tested this and i found an Problem with the String Parser DCCEXParcer.cpp. 
But back i want to Debug the Read from my Decoder, and Teensy alwas blocked after some reads.
You know the <D ACK ON> <R>   Debug test. I dont See the Printout on the serial monitor, but on mega its ok.
I try to get depper and i found this, maybee i do not Know how the Hash is working.

When i Send on Serial over the USB Port Teensy 4.1 the Command  <D AKC ON>   in the ParseD function called on 'D'  the Hash do not Match  with the "HASH_KEYWORD_ACK"  ! 
I try to check here:
`bool DCCEXParser::parseD(Print *stream, int params, int p[])
{
     Serial.println("Debug");

    if (params == 0)
        return false;

    bool onOff = (params > 0) && (p[1] == 1 || p[1] == HASH_KEYWORD_ON); // dont care if other stuff or missing... just means off
    
   Serial.println(p[0]);
   Serial.println(p[1]);
    Serial.println(p[2]);

    switch (p[0])
    {
    case HASH_KEYWORD_CABS: // <D CABS>
        DCC::displayCabList(stream);
        return true;

    case HASH_KEYWORD_RAM: // <D RAM>
        StringFormatter::send(stream, F("\nFree memory=%d\n"), freeMemory());
        break;

    case HASH_KEYWORD_ACK: // <D ACK ON/OFF> <D ACK [LIMIT|MIN|MAX] Value>
    Serial.println("Debug ACK");
`

Value on P[0] = 68649  and not 3113  so i never see the ACK debugging  when i  send <R> command.
Also Reading from DCC works but if i test form Com and JMRI it is not stable and Teensy  do  Hung.
If i do something  wrong let me know! 
I also worked with an Mega to Check the function and here its working.
I am very intrested on Teensy 4.x Platform pleae go on with that branch. I anyone can implemet NativeEthernet it was great.


Tanks.
Jake

Asbelos commented:
The numbers you get suggest that the teensy int is compiled as a 4 byte but the uno/mega would use 2 bytes. This would confuse the hash algorithm.

habazut commented:
You wrote:
> Value on P[0] = 68649 and not 3113 
What is an "int" on your platform? If it is not int16_t then this is not unexpected and the data type needs to be fixed at all places where the hash value is stored in an int.
 
            if (hot >= 'A' && hot <= 'Z')
            {
                // Since JMRI got modified to send keywords in some rare cases, we need this
                // Super Kluge to turn keywords into a hash value that can be recognised later
                runningValue = ((runningValue << 5) + runningValue) ^ hot;
                break;
            }
 
Harald.

mjs513 commented:
Just looking at this now.  

@Asbelos  is correct.  int on Teensy 4 and Teensy 3.x's are 4bytes not 2 bytes long.

Asbelos commented:
Just goes to show, porting takes more than just getting the compiler to work... There could be similar problems elsewhere bit i cant think of any. 
As an aside, sometimes it will be more efficient to use different int sizes, sometimes it will just be wrong... such is life when you dont have Java or C# to hide all this from you.

mjs513 commented:
`Just goes to show, porting takes more than just getting the compiler to work... There could be similar problems elsewhere bit i cant think of any.`

Totally agree. Ran into that problem a few times when porting libraries from some sensors like IMU's and Barometers.  Been awhile since I did that.  Unfortunately never tested the Diagnostic functions.  Did some additional testing with function buttons to test sounds with no issues and of course tested the DS52 track switches without problems.  

Was actually thinking if might be good to use int16's for all the parse functions.

habazut commented:
One way to deal with this is to never write "int" in your code any more but int16_t, uint16_t or int32_t and so on  what you need at that moment.

Should probably do that.

Harald.

mjs513 commented:
For now just updated parseD to use int16's.  Want to do some testing with other functions before making those.

jakespeed1311 commented:
Hello,
thanks for the explanation so far I had such a problem when I brought a program from the Mega to the Teensy.
But yes now that I have also sank it is clear.
Thanks also for the quick support, I like to keep testing and it doesn't matter if in such a complex code one or the other is found which helps us.

Therefore I am satisfied to have learned something and to have brought the branch further.

Greetings Jake

PS: Also Please go ahead with the branch and i hope to bring that in the Master.

jakespeed1311 commented:
Hi I tryed to Compile and thats not working. 
I am on Platformio all aktual Version, and i get an Convertion error  for int and int16_t 

Das Argument vom Typ ""int *"" ist mit dem Parameter vom Typ ""int16_t *"" inkompatibel. 

I removed the int16_t formthe parseD, so i wanted to define p to int16_t but i geht a red screen of errors.. bad Thening.

Then i take a look again to the hash Algorithmus you say and in the splitvalues the 
` int tunningValue  = 0 ;`

  changed to
`int16_t  tunningValue  = 0   ;`

and it seems thas will work.. and the Hash was ok 
the i tested with HASH_KEYWORD_PROGBOOST= -6353;  it have a negative value. 

So both -/+ values for tunningValue   are matching the devinition.
Meanwile i understand the <$ > command..

Is it ok to do it that way ?  

greedings Jake

PS: Test on mega also was ok too

jakespeed1311 commented:
Oh Sorry here the code change i misstiyped the var.

`      int DCCEXParser::splitValues(int result[MAX_PARAMS], const byte *cmd)

{
    byte state = 1;
    byte parameterCount = 0;
    int16_t runningValue = 0;
     `

mjs513 commented:
I think my first instinct was right - change all the int's to int16's in parser.  Ok now on my to do list.

jakespeed1311 commented:
HI,
sorry i don't know how to open an new thred. But i hope its ok to post hier.
Udpate do Ethernet for the Teensy 4.1 using NativeEthernet.

Just  Chane in EthernetInerface.h 
` #include "Ethernet.h"`

to

`    #include <NativeEthernet.h>         //TEENSY Ethernet Treiber `
`      #include <NativeEthernetUdp.h>  `


Litmited Testing with JMRI with the Throttle  (Drive an Cab with Nrr. 4) worked.
greedings
Jake

Asbelos commented:
Well done.  We need to do a bulk compare between your repo and ours so we can see all the differences you have added and make sure we can merge them without destabilizing everything.

jakespeed1311 commented:
Hello,
I still have one error that concerns the programming on the prog track. If I start the readout with <R> or <R 8 1 1> this can go 10-20 times well. But also abort at 1 bit while reading a byte.
At least I don't see a pattern, with the Mega it runs with the motor board and the decoder as expected.
Unfortunately I haven't found any clue why it behaves like this.
greedings
Jake


Here direct after the start reading out stop von CV8 bit 2 Teensy then will be frozen.


> DCC++ EX v3.0.3
> +++++ Ethernet Setup begin OK.
> LCD4:IP: 192.168.x.x
> 
> LCD5:Port:2560
> <iDCC-EX V-3.0.3 / TEENSY / STANDARD_MOTOR_SHIELD G-9db6d36>
> 
> LCD1:Ready
> <D ACK ON>
> Ack diag on
> <R 8 1 1>
> Auto Prog power on
> ACK baseline=18/53mA Threshold=38/113mA Duration: 2000us <= pulse <= 8500us
> V0 cv=8 bit=7
> NO-ACK after 145mS max=24/71mA pulse=0uS
> V1 cv=8 bit=7
> ACK after 33mS max=244/729mA pulse=5916uS
> V0 cv=8 bit=6
> ACK after 34mS max=263/786mA pulse=6148uS
> V0 cv=8 bit=5
> ACK after 34mS max=240/717mA pulse=5916uS
> V0 cv=8 bit=4
> NO-ACK after 145mS max=23/68mA pulse=0uS
> V0 cv=8 bit=3
> ACK after 36mS max=260/777mA pulse=6032uS
> V0 cv=8 bit=2
> 
###########################################

Here  Some reading until it stops

> DCC++ EX v3.0.3
> +++++ Ethernet Setup begin OK.
> LCD4:IP: 192.168.12.238
> 
> LCD5:Port:2560
> <iDCC-EX V-3.0.3 / TEENSY / STANDARD_MOTOR_SHIELD G-9db6d36>
> 
> LCD1:Ready
> <D ACK ON><R>
> Ack diag on
> 
> Auto Prog power on
> ACK baseline=8/23mA Threshold=28/83mA Duration: 2000us <= pulse <= 8500us
> V0 cv=1 bit=7
> ACK after 34mS max=229/684mA pulse=6032uS
> VB cv=19 value=0
> ACK after 37mS max=254/759mA pulse=6206uS
> V0 cv=29 bit=5
> ACK after 34mS max=236/705mA pulse=5858uS
> V0 cv=1 bit=6
> ACK after 34mS max=245/732mA pulse=6148uS
> V0 cv=1 bit=5
> ACK after 36mS max=240/717mA pulse=5916uS
> V0 cv=1 bit=4
> ACK after 35mS max=249/744mA pulse=6148uS
> V0 cv=1 bit=3
> ACK after 35mS max=240/717mA pulse=5917uS
> V0 cv=1 bit=2
> NO-ACK after 149mS max=18/53mA pulse=0uS
> V0 cv=1 bit=1
> ACK after 35mS max=237/708mA pulse=6206uS
> V0 cv=1 bit=0
> ACK after 36mS max=236/705mA pulse=5974uS
> VB cv=1 value=4
> ACK after 36mS max=254/759mA pulse=6148uSCallback
> 
> 
> Auto Prog power off
> Callback(4)
> <r 4><R 8 1 1>
> Auto Prog power on
> ACK baseline=12/35mA Threshold=32/95mA Duration: 2000us <= pulse <= 8500us
> V0 cv=8 bit=7
> NO-ACK after 145mS max=17/50mA pulse=0uS
> V1 cv=8 bit=7
> ACK after 35mS max=232/693mA pulse=6032uS
> V0 cv=8 bit=6
> ACK after 34mS max=243/726mA pulse=6148uS
> V0 cv=8 bit=5
> ACK after 34mS max=238/711mA pulse=5916uS
> V0 cv=8 bit=4
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V0 cv=8 bit=3
> ACK after 34mS max=246/735mA pulse=6148uS
> V0 cv=8 bit=2
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V0 cv=8 bit=1
> NO-ACK after 145mS max=17/50mA pulse=0uS
> V0 cv=8 bit=0
> NO-ACK after 145mS max=17/50mA pulse=0uS
> VB cv=8 value=151
> ACK after 34mS max=237/708mA pulse=5916uSCallback
> 
> 
> Auto Prog power off
> Callback(151)
> <r1|1|8 151><R 8 1 1>
> Auto Prog power on
> ACK baseline=12/35mA Threshold=32/95mA Duration: 2000us <= pulse <= 8500us
> V0 cv=8 bit=7
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V1 cv=8 bit=7
> ACK after 34mS max=234/699mA pulse=5858uS
> V0 cv=8 bit=6
> ACK after 34mS max=247/738mA pulse=6148uS
> V0 cv=8 bit=5
> ACK after 34mS max=237/708mA pulse=6032uS
> V0 cv=8 bit=4
> NO-ACK after 145mS max=17/50mA pulse=0uS
> V0 cv=8 bit=3
> ACK after 34mS max=237/708mA pulse=6032uS
> V0 cv=8 bit=2
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V0 cv=8 bit=1
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V0 cv=8 bit=0
> NO-ACK after 145mS max=18/53mA pulse=0uS
> VB cv=8 value=151
> ACK after 35mS max=251/750mA pulse=5916uSCallback
> 
> 
> Auto Prog power off
> Callback(151)
> <r1|1|8 151><R 8 1 1>
> Auto Prog power on
> ACK baseline=12/35mA Threshold=32/95mA Duration: 2000us <= pulse <= 8500us
> V0 cv=8 bit=7
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V1 cv=8 bit=7
> ACK after 34mS max=229/684mA pulse=5916uS
> V0 cv=8 bit=6
> ACK after 34mS max=250/747mA pulse=6148uS
> V0 cv=8 bit=5
> ACK after 34mS max=238/711mA pulse=6032uS
> V0 cv=8 bit=4
> NO-ACK after 145mS max=17/50mA pulse=0uS
> V0 cv=8 bit=3
> ACK after 34mS max=246/735mA pulse=6148uS
> V0 cv=8 bit=2
> NO-ACK after 145mS max=17/50mA pulse=0uS
> V0 cv=8 bit=1
> NO-ACK after 146mS max=17/50mA pulse=0uS
> V0 cv=8 bit=0
> NO-ACK after 145mS max=18/53mA pulse=0uS
> VB cv=8 value=151
> ACK after 34mS max=237/708mA pulse=5916uSCallback
> 
> 
> Auto Prog power off
> Callback(151)
> <r1|1|8 151><R 8 1 1>
> Auto Prog power on
> ACK baseline=11/32mA Threshold=31/92mA Duration: 2000us <= pulse <= 8500us
> V0 cv=8 bit=7
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V1 cv=8 bit=7
> ACK after 34mS max=231/690mA pulse=5800uS
> V0 cv=8 bit=6
> ACK after 34mS max=251/750mA pulse=6148uS
> V0 cv=8 bit=5
> ACK after 34mS max=234/699mA pulse=6032uS
> V0 cv=8 bit=4
> NO-ACK after 145mS max=18/53mA pulse=0uS
> V0 cv=8 bit=3
> ACK after 34mS max=242/723mA pulse=6148uS
> V0 cv=8 bit=2
> 
> 
> 
>

mjs513 commented:
@jakespeed1311 
Sorry was tied up with a few other projects.  But just made changes to DCCParser that changes all int's to int16t's but that also necessitated some changes to DCC.h/cpp as well. 

I never really played with <D> commands so if you want to download the update and give it a shot let me know.  Pretty much I am using the command window and web throttle to test the changes with.  I am not set up to do JMRI - another learning curve.

FrightRisk commented:
Need one of these to go with it and then program sounds for people who don't have a sound decoder. This has an audio board on it. https://www.pjrc.com/breakout-board-for-teensy-4-0/

Good work on all this. Are you still having the CV read issue?

jakespeed1311 commented:
Hello,
Yes i have JMRI to check the Ethernet and USB/Ser Commands.  And to come form the lovest level of test i go over the Serial Commands. I to this to find the CV read problem. And found the Hash problem.
1. Please add the Ethernet mod for Teensy 4.1 to the Branch. 

> `
#if  defined (TEENSYDUINO) 
 #if defined (ARDUINO_TEENSY41)
    #include <NativeEthernet.h>         //TEENSY Ethernet Treiber
    #include <NativeEthernetUdp.h>   
 #endif
``
#elif  ( !defined (TEENSYDUINO) && defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (ARDUINO_TEENSY41) )
 #include "Ethernet.h"

> #endif`





2.  The Debugging Command works now. <D ACK ON> So hash works. 
3. CV Reading is **not** working at all. I am Tested with <R> and <R 8 1 1> comand on console.    
    (Without ethernet Mod form Step 1)


Tank's for the Fixes comming so quick. I cross my fingers that we can figure out the cv read problem.
greedings
Jake

mjs513 commented:
@jakespeed1311 
2 things, easy one first.
Yes I can make that change - easy enough.  Was spending more time on the cv issue with D command R 8 1 1

Can I ask you a question - are you using the program track or main track.  I am using the Prog track?  Having a hard time tracing the hang.  Going to have to revert some of the int16's just to see what happens.  Think I am getting a handle on what links to what.  But this one is probably going to obvious - may be with setting up vargs.  Have to look at that next

jakespeed1311 commented:
hi,
i am using the  Programm Track.

mjs513 commented:
Ok all I think I traced it to an issue with the prog track.  When you do an "R 8 1 1" it winds up calling
```
      case V1:      // Issue validate bit=0 or bit=1  packet
        {
	  if (checkResets(blocking, RESET_MIN)) return; 
          if (Diag::ACK) DIAG(F("\nV%d cv=%d bit=%d"),opcode==V1, ackManagerCv,ackManagerBitNum); 
          byte instruction = VERIFY_BIT | (opcode==V0?BIT_OFF:BIT_ON) | ackManagerBitNum;
          byte message[] = {cv1(BIT_MANIPULATE, ackManagerCv), cv2(ackManagerCv), instruction };
		  Serial.println("DCCWaveform");
          DCCWaveform::progTrack.schedulePacket(message, sizeof(message), PROG_REPEATS);
		  Serial.println("DCCWaveform ack pending");
          DCCWaveform::progTrack.setAckPending(); 
        }
        break;
```

When it works it looks like its suppose to according the reference:

 ```
Auto Prog power on
ACK baseline=30/89mA Threshold=50/149mA Duration: 2000us <= pulse <= 8500us
V0 cv=8 bit=7DCCWaveform
DCCWaveform ack pending

NO-ACK after 145mS max=31/92mA pulse=0uS
V1 cv=8 bit=7DCCWaveform
DCCWaveform ack pending

ACK after 35mS max=171/511mA pulse=6612uS
V0 cv=8 bit=6DCCWaveform
DCCWaveform ack pending

ACK after 35mS max=271/810mA pulse=6960uS
V0 cv=8 bit=5DCCWaveform
DCCWaveform ack pending

ACK after 34mS max=193/577mA pulse=6438uS
V0 cv=8 bit=4DCCWaveform
DCCWaveform ack pending

ACK after 35mS max=165/493mA pulse=6728uS
V0 cv=8 bit=3DCCWaveform
DCCWaveform ack pending

NO-ACK after 145mS max=31/92mA pulse=0uS
V0 cv=8 bit=2DCCWaveform
DCCWaveform ack pending

NO-ACK after 145mS max=31/92mA pulse=0uS
V0 cv=8 bit=1DCCWaveform
DCCWaveform ack pending

ACK after 34mS max=162/484mA pulse=6380uS
V0 cv=8 bit=0DCCWaveform
DCCWaveform ack pending

NO-ACK after 145mS max=31/92mA pulse=0uS
VB cv=8 value=141
ACK after 34mS max=252/753mA pulse=6380uS
Auto Prog power off
Callback(141)
r1|1|8 141
```

When if fails it stuck it seems I think its still waiting or ack from the prog track:

```Auto Prog power on
ACK baseline=30/89mA Threshold=50/149mA Duration: 2000us <= pulse <= 8500us
V0 cv=8 bit=7DCCWaveform
DCCWaveform ack pending

NO-ACK after 145mS max=31/92mA pulse=0uS
V1 cv=8 bit=7DCCWaveform
DCCWaveform ack pending

ACK after 34mS max=174/520mA pulse=6148uS
V0 cv=8 bit=6DCCWaveform
DCCWaveform ack pending

ACK after 34mS max=165/493mA pulse=6438uS
V0 cv=8 bit=5DCCWaveform
DCCWaveform ack pending

ACK after 34mS max=162/484mA pulse=6844uS
V0 cv=8 bit=4DCCWaveform
DCCWaveform ack pending

ACK after 34mS max=168/502mA pulse=6264uS
**V0 cv=8 bit=3DCCWaveform
DCCWaveform ack pending**```

Any one have any ideas - power issue?

Asbelos commented:
The ack wait should timeout after the waveform has sent a given number of RESET packets.
You probably need to look in DCCWaveform checkAck()

mjs513 commented:
Traced through check and yes seems to timeout then the hang in some cases.
Here are the printf's I put in:
```
byte DCCWaveform::getAck() 
{
	  if(!ackPending) Serial.println("ACK PENDING FALSE: getACK");
      if (ackPending) return (2);  // still waiting
	  Serial.printf("ackDetected: %d\n",ackDetected);
      if (Diag::ACK) DIAG(F("\n%S after %dmS max=%d/%dmA pulse=%duS"),ackDetected?F("ACK"):F("NO-ACK"), ackCheckDuration,
           ackMaxCurrent,motorDriver->raw2mA(ackMaxCurrent), ackPulseDuration);
      if (ackDetected) return (1); // Yes we had an ack
      return(0);  // pending set off but not detected means no ACK.
}

void DCCWaveform::checkAck() {
    // This function operates in interrupt() time so must be fast and can't DIAG
    if (sentResetsSincePacket > 6) {  //ACK timeout
	Serial.println("Ack timeout");
```
On a good run:
```Auto Prog power on
ACK baseline=31/92mA Threshold=51/152mA Duration: 2000us <= pulse <= 8500us
V0 cv=8 bit=7DCCWaveform
DCCWaveform ack pending
Ack timeout
ACK PENDING FALSE: getACK
ackDetected: 0

NO-ACK after 145mS max=32/95mA pulse=0uS
V1 cv=8 bit=7DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 34mS max=193/577mA pulse=6380uS
V0 cv=8 bit=6DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 35mS max=164/490mA pulse=6670uS
V0 cv=8 bit=5DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 34mS max=267/798mA pulse=6032uS
V0 cv=8 bit=4DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 34mS max=176/526mA pulse=6380uS
V0 cv=8 bit=3DCCWaveform
DCCWaveform ack pending
Ack timeout
ACK PENDING FALSE: getACK
ackDetected: 0

NO-ACK after 145mS max=32/95mA pulse=0uS
V0 cv=8 bit=2DCCWaveform
DCCWaveform ack pending
Ack timeout
ACK PENDING FALSE: getACK
ackDetected: 0

NO-ACK after 145mS max=32/95mA pulse=0uS
V0 cv=8 bit=1DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 35mS max=162/484mA pulse=6844uS
V0 cv=8 bit=0DCCWaveform
DCCWaveform ack pending
Ack timeout
ACK PENDING FALSE: getACK
ackDetected: 0

NO-ACK after 145mS max=32/95mA pulse=0uS
VB cv=8 value=141ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 34mS max=162/484mA pulse=6728uS
Auto Prog power off
Callback(141)
<r1|1|8 141>
```
On a couple runs after:
```
ACK baseline=30/89mA Threshold=50/149mA Duration: 2000us <= pulse <= 8500us
V0 cv=8 bit=7DCCWaveform
DCCWaveform ack pending
Ack timeout
ACK PENDING FALSE: getACK
ackDetected: 0

NO-ACK after 145mS max=32/95mA pulse=0uS
V1 cv=8 bit=7DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 35mS max=173/517mA pulse=6496uS
V0 cv=8 bit=6DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 35mS max=167/499mA pulse=6844uS
V0 cv=8 bit=5DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 34mS max=197/589mA pulse=6264uS
V0 cv=8 bit=4DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACK
ackDetected: 1

ACK after 35mS max=166/496mA pulse=6612uS
V0 cv=8 bit=3DCCWaveform
DCCWaveform ack pending
Ack timeout
```

jakespeed1311 commented:
hi,
yes, thats what i see too. But can it be hat the reset counter will not be come back right time or value ?
Sorry will do more but can't.
greedings
Jake

Asbelos commented:
Its bizarre because you have run this many times ok but it seems to stop in the same place although there is no difference between the bit checks for any bit of any cv. 

Couple of questions...
1. Are you running the commands over the Ethernet or Wifi commecrion or just Usb ? This changes the way the code waits!
2. If the system stops like this does it still respond to <s> commands or nothing. 
If it responds to nothing AND  the command came over usb then its Not waiting for the ack... but maybe something else like a main track current poll has crashed the system or corrupted memory.

Asbelos commented:
Does this happen a certain amount of time after a reboot maybe?

mjs513 commented:
@Asbelos 
Its really bizarre.  As for my testing I am using just USB (Serial Monitor - Arduino IDE).  After the last Ack timeout if I try to enter another command there is no response.

As to your last question - I tested the loco on the main track with throttle app and worked no problem.  Then I shut that down and moved it to a prog track and ran the test same thing happened.

I did test the code on a Mega and a Teensy 3.5 without the problem showing up.  Going to work on it more later today.

jakespeed1311 commented:
hi,
iam use only the usb connection. And i geht no responce after i send an new comand. I don't see an time isue from reset of the board to the stuck. So i have reset the board and cap driving runs one hour fine. Then i go to prog track and it works serveral times, but then ist stops. I  i reset and got to prog same behavior.
hope this helps
jake

Asbelos commented:
This is starting to look like a RAM leak or stack corruption.. ... did you manage to.port the freemem warning in the .ino? 

Did you add any diagnostics that might use the  String class?

mjs513 commented:
Will take a look at freemem - think I have something that works with the
T4x.  But will take a look at that next.

No added diagnostics - just added Serial.printf's to dump debug info.
Tried to keep it to a minimum.

On Wed, Feb 24, 2021 at 9:53 AM Asbelos <notifications@github.com> wrote:

> This is starting to look like a RAM leak or stack corruption.. ... did you
> manage to.port the freemem warning in the .ino?
>
> Did you add any diagnostics that might use the String class?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/DCC-EX/CommandStation-EX/pull/125#issuecomment-785130887>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABI6DRJX5QTXTQQVROJXC4LTAUHGRANCNFSM4W3QXYLQ>
> .
>

jakespeed1311 commented:
No, short answer.

mjs513 commented:
Ok I am absolutely stuck.
1,  Added the freemem equivalent for T4.x's:
```Free memory=437452
Heap memory=511664
```
2,  I added one more print statement:
```ACK after 32mS max=161/481mA pulse=6264uS
V0 cv=8 bit=2DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACKackDetected: 1

ACK after 32mS max=163/487mA pulse=6380uS
V0 cv=8 bit=1DCCWaveform
DCCWaveform ack pending
ACK PENDING FALSE: getACKackDetected: 1

ACK after 32mS max=153/457mA pulse=6264uS
V0 cv=8 bit=0DCCWaveform
DCCWaveform ack pending
Ack timeout
ackPending: 0
```
On the off chance that maybe need digitalwrites/reads to be faster since I am running at 600Mhz I edited the motordriver.cpp:
```#elif defined(TEENSYDUINO)
    #define WritePin digitalWriteFast
    #define ReadPin digitalReadFast
#else
......
```

mjs513 commented:
Just posted the issue on the PJRC forum - maybe somebody over there has an idea:
https://forum.pjrc.com/threads/66384-Help!-Program-works-on-T3-5-but-fails-on-T4-x

jakespeed1311 commented:
HI,
i will try the last Teensyduino beta maybee it is there an hint.

mjs513 commented:
@jakespeed1311 
Just as an FYI I am using the latest, Beta7 :)  Right now there is a lot of activity on the forum link I posted if you want to read.

jakespeed1311 commented:
hmm ok, yes i read in the Forum.
But i can't follow that level of Diskussion.

jakespeed1311 commented:
i coud try to setup an ttl to usb converter con  serialX to move from away from  usb port

mjs513 commented:
Don't think thats necessary - i have a Logic Analyzer tied in watching the signal.  And if using a T4.1 the mods that luni showed for dual serial works just not a T3.5 or T3.6.

You don't have a T3.5 you can test with by any chance.  Seems like a T3.5/T.36 works.

jakespeed1311 commented:
hi,
i have only Teensy 4.1 now.

mjs513 commented:
Ok making some progress - after instrumenting waveform and parser with digitaltoggles and my logic ananlyzer(LA) the problem appears to be in checkPowerOverload for the program track.  If I comment the check out in the loop (DCCWaveform I can send a <R 8 1 1> all day.

So next step is to analyze what's going on with that function.

mjs513 commented:
Tracked it down to one call in DCCWaveform::checkPowerOverload.
`      lastCurrent = motorDriver->getCurrentRaw();`

If I comment that line out I can send <R 8 1 1> all day!

And before you say that because I am using analogRead and not AnalogReadFast.  I used to LA on a Mega and the Teensy 4.1 to get the timing for both (int adc = ...):
```
Mega - 22.2 microseconds
T4.1   - 17.2 microseconds
```

so now at a total loss... any suggestions?

Asbelos commented:
That suggests to me that analogRead on a teensy is not interrupt safe.
You may have to wrap nointerruot around the call from the loop but not the call from rhe interrupt looking for acks.

mjs513 commented:
Morning @Asbelos 
Just got up and took your suggestion so my loop function in DCCWaveform now looks like this:
```
void DCCWaveform::loop() {
  mainTrack.checkPowerOverload();
  noInterrupts();
  progTrack.checkPowerOverload();
  interrupts();
}
```
and spent the last 5 minutes sending many many "R 8 1 1" s.  Not going to dump the serial monitor out here but no hangs.  Posted the function for @jakespeed1311.  Now to clean up.

jakespeed1311 commented:
HI,
Great work,, i check now !

jakespeed1311 commented:
HI,
Fast Answer YOU DIT IT it works now.   In German i wold say  " Laeuft wie ein Uhrwerk"      "It runs like clockwork" 

No hang after 40 time  <R 8 1 1> call. 

Now i test with jmri over usv serial to read out geh complete ESU decoder.
I know that is maximum stress so it reads all CV in the table. I try READ Full Sheet on the CV's tab all 1024 !
Readmode Direkt Byte -   I Stoped on CV 200  NO read fail no stop.
Readmode Direkt Bit     - I stoped on CV  165 also no Problem.

I am very luky about that ;-)
greedings
Jake

PS:
My next steps is to Progamm on the Service Trak and to Programm on Main. 
So i saw Railcom Reading is not yet  implemented, but thats no problem. I but it was cool wenn that also will come to main.

Meanwile readed CV 511 out..

mjs513 commented:
I pushed up that change so the PR is updated.

Also tested the changes on Mega 2560, T3.2, T3.5, T3.6, T4.0 and T4.1 => all work.

@jakespeed1311 - let me know if you find anything during testing either on the program track or the main track.

jakespeed1311 commented:
Yes, did you push the ethernet change ?

mjs513 commented:
Oh shoot - thanks for reminding me - will do that with the next update.   Need my cup of coffee first :)

mjs513 commented:
`#if defined (TEENSYDUINO)
#if defined (ARDUINO_TEENSY41)
#include <NativeEthernet.h> //TEENSY Ethernet Treiber
#include <NativeEthernetUdp.h>
#endif
``
#elif ( !defined (TEENSYDUINO) && defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (ARDUINO_TEENSY41) )
#include "Ethernet.h"

#endif``

@jakespeed1311 - what function does this go in - can't seem to find it right now.

jakespeed1311 commented:
HI,
it is in Ethernetinterface.h line 31  round.
Sorry I cant get copy paste code between the code Quouts (Inser Code) it alwas get wrong.

`#ifndef EthernetInterface_h
#define EthernetInterface_h

#include "DCCEXParser.h"
#include "MemStream.h"
#include <Arduino.h>
#include <avr/pgmspace.h>

#if  defined (TEENSYDUINO) 
 #if defined (ARDUINO_TEENSY41)
    #include <NativeEthernet.h>         //TEENSY Ethernet Treiber
    #include <NativeEthernetUdp.h>   
 #endif


#elif  ( !defined (TEENSYDUINO) && defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO) )
 #include "Ethernet.h"
#endif

//#include <NativeEthernet.h>
//#include <Ethernet.h>
#include "RingStream.h"

/**
 * @brief Network Configuration
 * `

Some usefull info i hope
But native ethernet is function compatible with the wiz ethernet.h
i think also in Teensyduino up on last version Beta Natice Ethernet it is included.
You need if not in 1.54
[https://github.com/vjmuzik/NativeEthernet](url)
and the real hardwar driver
[https://github.com/vjmuzik/FNET](url)

Tanks to vjmuzik for that.

> ust change <Ethernet.h> to <NativeEthernet.h> to use this library and the rest is taken care of. The API is a direct replacement for the normal Ethernet.h library aside from sketches that use direct calls to w5100.h

I hope this it the anser to the question.
greedings
Jake

jakespeed1311 commented:
Oh yes now i geht it 
```
#ifndef EthernetInterface_h
#define EthernetInterface_h
#include "DCCEXParser.h"
#include "MemStream.h"
#include <Arduino.h>
#include <avr/pgmspace.h>
#if  defined (TEENSYDUINO) 
 #if defined (ARDUINO_TEENSY41)
    #include <NativeEthernet.h>         //TEENSY Ethernet Treiber
    #include <NativeEthernetUdp.h>   
 #endif


#elif  ( !defined (TEENSYDUINO) && defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO) )
 #include "Ethernet.h"
#endif

//#include <NativeEthernet.h>
//#include <Ethernet.h>
#include "RingStream.h"

/**
 * @brief Network Configuration
 * 
```

jakespeed1311 commented:
Iforgot  we need to changein 

> defines.h

```
#if ENABLE_WIFI && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (TEENSYDUINO))
#define WIFI_ON true
#else
#define WIFI_ON false
#endif

#if ENABLE_ETHERNET && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (TEENSYDUINO))
#define ETHERNET_ON true
#else
#define ETHERNET_ON false
#endif
#if ENABLE_WIFI && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (TEENSYDUINO))
#define WIFI_ON true
#else
#define WIFI_ON false
#endif

#if ENABLE_ETHERNET && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (TEENSYDUINO))
#define ETHERNET_ON true
#else
#define ETHERNET_ON false
#endif
```

mjs513 commented:
@jakespeed1311 
Think your ifdef's can be boiled down to:
```
#if defined (ARDUINO_TEENSY41)
 #include <NativeEthernet.h>         //TEENSY Ethernet Treiber
 #include <NativeEthernetUdp.h>   
#else
 #include "Ethernet.h"
#endif
```

jakespeed1311 commented:
yes you are right!

mjs513 commented:
@jakespeed1311 
Sorry for the delay in getting this incorporated, with the issue we were I forgot all about it, but its there now.  Let me know how it works out in your testing.

@Asbelos 
I am in the process of what the issue is with analogRead for the T4.x.  Running a simple sketch with analogRead in the interrupt isn't showing any issues.    A quick question, do all of you have a simple test sketch just to generate the DCC waveform?  I can't seem to find where state is set in the interrupt function in DCCWaveform?

jakespeed1311 commented:
Hi,
there is an Type in defines.h
#if ENABLE_WIFI && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)) || defined(Teensyduino)
The one ) after Sam_Zero must be the last char in the line.

Here:
#if ENABLE_ETHERNET && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (TEENSYDUINO))
and
#if ENABLE_WIFI && (defined(ARDUINO_AVR_MEGA) || defined(ARDUINO_AVR_MEGA2560) || defined(ARDUINO_SAMD_ZERO)||defined (TEENSYDUINO))

habazut commented:
I guess that the problem is not in the analogRead but the integer lastCurrent. lastCurrent is several bytes wide so setting it is not an atomic operation. If you get an interrupt in the middle of setting lastCurrent then the value can be aything. As an int is 4 bytes on the Teensy and faster clock I guess that the chance that it is corrupted is much higher.

I would make 2 changes which you might want to test:

1. Change int to int16_t where current and friends is used
2. Put he nointerrupt only around the assignement like this:


    case POWERMODE::ON:
      // Check current
      noInterrupts();
      lastCurrent = motorDriver->getCurrentRaw();
      interrupts();
      if (lastCurrent <= tripValue) {

If my guesses are right that should do it. I can not test as I don't have the HW.

Dann woll mer mal Daumen halten,
Harald.

mjs513 commented:
@habazut 
Gave your suggestion a try and it appears to work as well on the T4.1

I did try the int16 change but that didn't work.  turning interrupts on and off seems to do the trick.

jakespeed1311 commented:
;-)

Asbelos commented:
Pleasd bear in mind the imminent replacement of the waveform generator and timing stack... 
I strongly suggest you catalogue the kinds of changes you have had to make for teensy and prepare to re-craft hem jf necessary

Asbelos commented:
Sorry i fidnt mean tl close that... fonger yrouble pn phonr

mjs513 commented:
Know what you mean about working on your phone.  Unless @jakespeed1311 finds another something else think it pretty much done for now.

habazut commented:
I did write the following in an email but it did not seem to get here....

> I did try the int16 change but that didn't work.  turning interrupts on and off seems to do the trick.

Probably can do it yet shorter:

    case POWERMODE::ON:
      // Check current
      { 
        int c = motorDriver->getCurrentRaw();
        noInterrupts;
        lastCurrent = c;
        interrupts;
      }

Can you try that as that would show us if the problem is in the
analogRead or in the assignement. We want to block interrupts as short
as possible.

It probably does not matter if int32_t or int16_t
as both are > 1 byte and not atomic.

getCurrentRaw() should probably be inline.

Harald.

mjs513 commented:
Gave your changes a try but still got stuck, same bad behavior as before.

Now to test analogRead what I did was:
```
case POWERMODE::ON:
  // Check current
  { 
    //int c = motorDriver->getCurrentRaw();
    noInterrupts;
    lastCurrent = analogRead(A1);
    interrupts;
  }
```
That worked without a problem.    Seems the problem is the call off to getCurrentRaw.

Now with all that said.  I added this to DCCWaveform.h based on your suggestion:
```
    inline int  getCurrentRaw() {
	  if (powerMode == POWERMODE::ON)
	      return motorDriver->getCurrentRaw();
	  return 0;
    }	
```
Put all the int16's back to int's in DCCWaveform and commented out NoInterrupts/interrupts and....
**Volia - that worked perfectly - no hangs on the T4.x's**

Now have to go test with T3.x's to see what happens with those.

mjs513 commented:
@habazut 
Ok - i messed up - I edited the file but forgot to save the changes so --- bottom line - even with the inline still need the interrupt for the T4.  

To avoid any issue on the mega or avr easy enough to put it in a ifdef statement......

mjs513 commented:
@Asbelos 
```
Pleasd bear in mind the imminent replacement of the waveform generator and timing stack...
I strongly suggest you catalogue the kinds of changes you have had to make for teensy and prepare to re-craft hem jf necessary
```

Any idea when this is going to happen?

Asbelos commented:
The master has now been updated with a vast number of changes aimed at performance, features and portability.

Please rea-evaluate on your Teensy processors and fell free to ask questions or suggest updates, but please do his on our Discord group in the Hardware #arduino  channel.

mjs513 commented:
Thanks will do.

mjs513 commented:
Hi Jake
Going to work on it over the several days.  Right in the middle of another issue that I am trying to resolve.  Once that's done then on to DCC-EX.

Mike

jakespeed1311 commented:
Ok, tanks Mike

jakespeed1311 commented:
Hi Mike,
I hope you get on with your problem,
It would be good if Teensy comes in the master, I am just about to include loconet.
Greetings
Jake  
PS: Sorry this may come across as impatient, but I am really excited for this.

mjs513 commented:
Hi Jake
Sorry its been taking so long but other things on the list to take of as well, but hasn't been far from my mind.  Every time I think I am going to get back to it something else happens.

Anyway its funny you bring this up now, I just started getting back to it this morning.  Some things changed in new version and have to make sure they are still compatible with Teensy.  So far so good.  Hopefully today I will have the Timer code updated so I can at least test that piece then on to see if I still have to make changes again to Diag stuff again.

Maybe by the weekend if something else doesn't come up.

Mike

Asbelos commented:
Mike, 
If you use git and do your dev in a teensy branch then I can keep an eye on what you change so I'll be able to help or spot anything that might cause incompatibilities when merging.

mjs513 commented:
@Asbelos
Thanks but I think I got it working.  Tested the waveform and its looking good.  LA shows 58us.  Retested the issue with <D ACK..> commands and they are working as well.  Had to check a few things with compatibility.  Just took a little time between other projects and issues.  Didnt add the PWM pin stuff yet but that will come after the rest is checked out.  Want to check out LCD and main track functions - haven't down that yet.

@jakespeed1311 - before I do another PR here is my updated branch: https://github.com/mjs513/CommandStation-EX/tree/Teensy-newVersion.  Yeah I know I am original in naming conventions.  Can you check out the ethernet stuff and anything else you can think of.

Mike

Asbelos commented:
I put a comment on your latest checkin.. I hope it gets to you. Are you on our discord?

mjs513 commented:
Think I messed something up with my github.  I saw what you sent but didn't know how to get it into the system since it wasn't attached to a branch so I just closed it.  The TeensyBranch is the old version - Teensy-newVersion is what I was working on.

Yes I am on discord.  Forgot my user name :)  I just login and it shows up :)

Asbelos commented:
Well login discord and friend UKBloke (AKA Chris)  ... its easier than email
